Fire event on rflink sensor update

akasma74 opened the pull request:
## Description:
For long time I wanted to have a way to tell whether a sensor sends its readings or it's just broke/run out of batteries.
There was no way to find out when the last reading was received by rflink component and without that it was impossible to tell whether the sensor sends the same value over a long period of time OR it's out of order and is sending nothing at all (because in both cases HA does not update its state).

That's why I thought it would be useful to have a fire_event option (exactly as in `rflink.switch`) for sensors so it can be used as a heartbeat.

I'm currently use customised rflink  component in my local Hass.io environment and it does the job.

**Related issue (if applicable):** fixes #<home-assistant issue number goes here>

**Pull request in [home-assistant.io](https://github.com/home-assistant/home-assistant.io) with documentation (if applicable):** home-assistant/home-assistant.io#<home-assistant.io PR number goes here>

## Example entry for `configuration.yaml` (if applicable):
```yaml

```

## Checklist:
  - [x] The code change is tested and works locally.
  - [ ] Local tests pass with `tox`. **Your PR cannot be merged unless tests pass**
  - [ ] There is no commented out code in this PR.

If user exposed functionality or configuration variables are added/changed:
  - [ ] Documentation added/updated in [home-assistant.io](https://github.com/home-assistant/home-assistant.io)

If the code communicates with devices, web services, or third-party tools:
  - [ ] New dependencies have been added to the `REQUIREMENTS` variable ([example][ex-requir]).
  - [ ] New dependencies are only imported inside functions that use them ([example][ex-import]).
  - [ ] New or updated dependencies have been added to `requirements_all.txt` by running `script/gen_requirements_all.py`.
  - [ ] New files were added to `.coveragerc`.

If the code does not interact with devices:
  - [ ] Tests have been added to verify that the new code works.

[ex-requir]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L14
[ex-import]: https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/keyboard/__init__.py#L23

akasma74 committed:
Update __init__.py

Added firing event on sensor readings arrival if the fire_event flag is set for that sensor

akasma74 committed:
Update sensor.py

added fire_event option to the sensor configuration options

homeassistant commented:
Hi @akasma74, 

It seems you haven't yet signed a CLA. Please do so [here](https://home-assistant.io/developers/cla_sign_start/?pr=home-assistant&#x2F;home-assistant#22619).

Once you do that we will be able to review and accept this pull request.

Thanks!

akasma74 committed:
fire_event option added to platform schema

fixed "line too long"s

javicalle commented:
I understand the purpose of the request (I have point myself to it a lot of times) but IMHO I think it's not an RFLink problem, and not must be point to a RFLink solution.
IMHO it should be the behavior of the `State` class that should cover this need for **any platform**. Device's `state` have an `last_updated` (diferent from `last_changed`) that could track when a sensor reading arrives.
That could be a homogeneous solution for all the devices that would avoid having to make adaptations in each platform.

Obviously this is a "core" behavior that should not be tackled lightly, but (although I perfectly understand the need) I do not feel comfortable looking for a particular solution.

akasma74 commented:
> IMHO it should be the behavior of the `State` class that should cover this need for **any platform**. Device's `state` have an `last_updated` (diferent from `last_changed`) that could track when a sensor reading arrives.
> That could be a homogeneous solution for all the devices that would avoid having to make adaptations in each platform.

That sounds great, but can you predict roughly in what year it might be implemented?
I made this change because I had a need for it for about a year and could see no standard HA solution. When I realised that switch.rflink has that functionality, I thought it's natural to extend another rflink platform with such an important feature. Yes, my PR does not make the whole HA ecosystem better, but it's pretty "homogenous" and can really help people like me to achieve what they need using standard components and not messing about with custom components etc.

> Obviously this is a "core" behavior that should not be tackled lightly, but (although I perfectly understand the need) I do not feel comfortable looking for a particular solution.

And what are you more comfortable looking for in this particular department?
Is there a real alternative that can be available to everyone in the near future?

MartinHjelmare commented on homeassistant/components/rflink/__init__.py:
```diff
@@ -50,6 +50,7 @@
 CONNECTION_TIMEOUT = 10
 
 EVENT_BUTTON_PRESSED = 'button_pressed'
+EVENT_SENSOR_READINGS = 'sensor_readings'
```
Event names must be prefixed with the integration name. In this case `rflink_`.

emontnemery commented:
@Akasma74 have a look at `off_delay_listener` in rflink/binary_sensor.py something similar can be used to set the sensor unavailable if no data is received for a certain time. Also, `last_updated` which @javicalle points to is already available.

akasma74 commented:
> @akasma74 have a look at `off_delay_listener` in rflink/binary_sensor.py something similar can be used to set the sensor unavailable if no data is received for a certain time. 

Well, it implies we need expiry_after option to be added to rflink.sensor, isn't it?
And even if we have that, it's not as flexible as an event because it's fixed in configuration and there is no way to change it dynamically, am I wrong?
Having a fire_event option allows to do much more.
Ideally, though, we need both as not everyone want and need to mess with events or their goal is easily achieved using expiry_after.

> Also, `last_updated` which @javicalle points to is already available.

Ok, I was probably too quick. Let me explain.
According to [state object docs](https://www.home-assistant.io/docs/configuration/state_object/) on last_updated, "writing the exact same state including attributes will not result in this field being updated". That makes impossible to distinguish between "sensor sends the same temperature for the last hour" and "sensor's batteries died an hour ago". Same goes to last_updated, it's a design issue.

And as a proof that last_updated of sensor.rflink is not fit to purpose I added this code to the automation that receives a reading event and toggles input_boolean that represents heartbeat:
```
          {% set sensor_object_id = trigger.to_state.object_id %}
          {{ "{}.last_updated={}".format( sensor_object_id, states.sensor[sensor_object_id].last_changed ) }}
          {{ "{}.last_updated={}".format( sensor_object_id, states.sensor[sensor_object_id].last_updated ) }}
          {% set heartbeat_object_id = 'sensor_heartbeat_' + sensor_object_id %}
          {{ "{}.last_changed={}".format( heartbeat_object_id, states.input_boolean[heartbeat_object_id].last_changed ) }}
``` 

here is its output:

> ground_floor_kitchen_temperature.last_changed=2019-04-02 07:41:11.434395+00:00
> ground_floor_kitchen_temperature.last_updated=2019-04-02 07:41:11.434395+00:00
> sensor_heartbeat_ground_floor_kitchen_temperature.last_changed=2019-04-02 07:50:24.416958+00:00
> 

By the way, do you know why all that last_xxx things are now different to {{ now() }} ?
Because it gives me 

> 2019-04-02 08:52:09.579233+01:00

an hour difference. I noticed that as we changed to Summer time..

javicalle commented:
> According to state object docs on last_updated, "writing the exact same state including attributes will not result in this field being updated". That makes impossible to distinguish between "sensor sends the same temperature for the last hour" and "sensor's batteries died an hour ago". Same goes to last_updated, it's a design issue.

This is the current behavior for any device. It's strange to me, but...

I have not tested it personally, but would not it be possible to have this behavior with the `force_update` attribute?
* https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity.py

akasma74 commented:
> > According to state object docs on last_updated, "writing the exact same state including attributes will not result in this field being updated". That makes impossible to distinguish between "sensor sends the same temperature for the last hour" and "sensor's batteries died an hour ago". Same goes to last_updated, it's a design issue.
> 
> This is the current behavior for any device. It's strange to me, but...

first of all, I presume that's because they didn't want to update states on every single event if there is nothing really to update - just imagine thousands of entities overwriting the same values, it'd be a killer.
on the other hand, that approach gives us the above mentioned problem (and even some more, all related to this design) and it needs to be addressed.

> I have not tested it personally, but would not it be possible to have this behavior with the `force_update` attribute?
> 
> * https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/helpers/entity.py

could you elaborate?

javicalle commented:
> could you elaborate?

I'll try.

According to the code comment:

>         """Return True if state updates should be forced.
>        If True, a state change will be triggered anytime the state property is
>        updated, not just when the value changes.
>        """

This should force device's state update at every reading.

I believe that `rflink.binary_sensor` has the functionality implemented through the device's configuration:
- https://www.home-assistant.io/components/binary_sensor.rflink/#force_update
- https://github.com/home-assistant/home-assistant/blob/16e0953f267f6b24cbbfebee282a48cb542f7bb4/homeassistant/components/rflink/binary_sensor.py#L102

javicalle commented:
> I presume that's because they didn't want to update states on every single event if there is nothing really to update - just imagine thousands of entities overwriting the same values, it'd be a killer

I understand that is not wanted to fire a change of state in each reading, but (as you has proved) in the current implementation `last_updated` attribute not makes sense (maybe at reboots?).
As I see it, with 0 knowledge from core components and apart from the implications that may have, `last_updated` must reflect the last reading time, and `last_changed` the last state changed fired event time.

akasma74 commented:
> > I presume that's because they didn't want to update states on every single event if there is nothing really to update - just imagine thousands of entities overwriting the same values, it'd be a killer
> 
> I understand that is not wanted to fire a change of state in each reading, but (as you has proved) in the current implementation `last_updated` attribute not makes sense (maybe at reboots?).
> As I see it, with 0 knowledge from core components and apart from the implications that may have, `last_updated` must reflect the last reading time, and `last_changed` the last state changed fired event time.

Err.. not quite. If I get it right these two serve 2 different roles.
1. Entity has its state (on/off/available etc). It's a major thing and last_changed reflects when it changed, i.e it went from off to on. If it was on and HA received "turn it on" command, there is no point to update it as it's already on and if later someone enquires "when it was on", there is no point to mention that on command, is it? That's why last_changed won't be updated if the state is the same as current one.
2. Entity may have plenty of attributes like name, icon etc. It's just extra bits and bobs, but you can also track their changes if you wish. Again, if we say "set icon mdi:question" and the entity already has that icon set, what's the point of changing it, updating HA state machine etc? Just move on, everything's done already.

You might argue that it's better to update those last_xxx even without writing anything to the state machine - maybe, I don't know all the implications exactly because of "0 knowledge from core components" ;)

I'm trying to add that force_update option to rflink.sensor, but it does not let me, validation says Invalid option.. :\

javicalle commented:
I'm not sure if we are pointing the same or not. Anyway I'm going to make some tests before reply :-?

> I'm trying to add that force_update option to rflink.sensor, but it does not let me, validation says Invalid option.. :\

`force_update` is not implemented for `rflink.sensor`, but I think that can be easily copied from `rflink.binary_sensor` and maybe would make the work.

akasma74 commented:
> I'm not sure if we are pointing the same or not. Anyway I'm going to make some tests before reply :-?
> 
> > I'm trying to add that force_update option to rflink.sensor, but it does not let me, validation says Invalid option.. :\
> 
> `force_update` is not implemented for `rflink.sensor`, but I think that can be easily copied from `rflink.binary_sensor` and maybe would make the work.

mate, I looked at binary_sensor's force_update feature and implemented it in sensor.
actually, I currently have both force_update and expire_after (as in MQTT sensor) and I think it's a great combination, works like a treat.

emontnemery commented:
> For long time I wanted to have a way to tell whether a sensor sends its readings or it's just broke/run out of batteries.

I still think the best way to achieve this goal is to implement something similar to `off_delay_listener` in rflink/binary_sensor.py which would set the device to unavailable. All other solutions require a lot of manual configurations with scripting etc.

akasma74 commented:
> > For long time I wanted to have a way to tell whether a sensor sends its readings or it's just broke/run out of batteries.
> 
> I still think the best way to achieve this goal is to implement something similar to `off_delay_listener` in rflink/binary_sensor.py which would set the device to unavailable. All other solutions require a lot of manual configurations with scripting etc.

I've abandoned fire_event for epxire_after and force_update, please read last bit of my [comment](https://github.com/home-assistant/home-assistant/pull/22619#issuecomment-478947178)

emontnemery commented:
Another option is to set a last_seen attribute or similar on the sensor each time we get data from RFLink

akasma74 commented:
But what's wrong with force_update/expire_after?

balloob commented:
We should not fire an event when a sensor is read. We have state changed events which contain last_updated.

emontnemery commented:
> But what's wrong with force_update/expire_after

Nothing at all, I just didn't read your comment carefully :)
